File: scala/src/test/scala/demo/styles/OutputBasedTests.scala
package demo.styles

import demo.customer.{PriceEngine, Product}
import org.scalatest.flatspec.AnyFlatSpec

class OutputBasedTests extends AnyFlatSpec {
  it should "calculate the discount of 2 products" in {
    val product1 = Product("Kaamelott")
    val product2 = Product("Free Guy")

    // Call on the SUT (here PriceEngine)
    // No side effects -> Pure function
    val discount = PriceEngine.calculateDiscount(product1, product2)

    assert(0.02 == discount)
  }
}

File: php/tests/Styles/OutputBasedTests.php
<?php

namespace Styles;

use Com\CleanTests\Customer\PriceEngine;
use Com\CleanTests\Customer\Product;
use PHPUnit\Framework\TestCase;
use function PHPUnit\Framework\assertEquals;

class OutputBasedTests extends TestCase
{
    public function testDiscountOf2ProductsShouldBe2Percent(): void
    {
        $product1 = new Product("Kaamelott");
        $product2 = new Product("Free Guy");

        // Call on the SUT (here PriceEngine)
        // No side effects -> Pure function
        $priceEngine = new PriceEngine();
        $discount = $priceEngine->calculateDiscount([$product1, $product2]);

        assertEquals(0.02, $discount);
    }
}